adapters/zio-http/src/main/scala/caliban/ZHTTPAdapter.scala

```diff
@@ -0,0 +1,240 @@
+                               .foldLeft(query)((q, _) => q.withFederatedTracing)
+          resp            <- executeToJson(interpreter, queryWithTracing, skipValidation, enableIntrospection, queryExecution)
+        } yield Response.jsonString(resp.toString)).handleHTTPError
+      case req @ Method.GET -> _  =>
+        val params = List("query", "operationName", "variables", "extensions")
+          .collect({ case k =>
+            k -> req.url.queryParams.get(k).flatMap(_.headOption).getOrElse("")
+          })
+          .toMap
+
+        (for {
+          query <- ZIO.fromEither(decode[GraphQLRequest](params.asJson.noSpaces))
+          resp  <- executeToJson(interpreter, query, skipValidation, enableIntrospection, queryExecution)
+        } yield Response.jsonString(resp.toString())).handleHTTPError
+
+      case _ @Method.OPTIONS -> _ =>
+        ZIO.succeed(Response.http(Status.NO_CONTENT))
+    }
+
+  def makeWebSocketService[R <: Has[_], E](
+    interpreter: GraphQLInterpreter[R, E],
+    skipValidation: Boolean = false,
+    enableIntrospection: Boolean = true,
+    keepAliveTime: Option[Duration] = None,
+    queryExecution: QueryExecution = QueryExecution.Parallel
+  ) =
+    HttpApp.responseM(
+      for {
+        ref <- Ref.make(Map.empty[String, Promise[Any, Unit]])
+      } yield Response.socket(
+        socketHandler(
+          ref,
+          interpreter,
+          skipValidation,
+          enableIntrospection,
+          keepAliveTime,
+          queryExecution
+        )
+      )
+    )
+
+  private def socketHandler[R <: Has[_], E](
+    subscriptions: Ref[Map[String, Promise[Any, Unit]]],
+    interpreter: GraphQLInterpreter[R, E],
+    skipValidation: Boolean,
+    enableIntrospection: Boolean,
+    keepAliveTime: Option[Duration],
+    queryExecution: QueryExecution
+  ): SocketApp[R with Clock, E] = {
+    val routes = Socket.collect[WebSocketFrame] { case Text(text) =>
+      ZStream
+        .fromEffect(ZIO.fromEither(decode[GraphQLWSRequest](text)))
+        .collect({
+          case GraphQLWSRequest("connection_init", _, _)      => connectionAck ++ keepAlive(keepAliveTime)
+          case GraphQLWSRequest("connection_terminate", _, _) => close
+          case GraphQLWSRequest("start", id, payload)         =>
+            payload match {
+              case Some(req) =>
+                generateGraphQLResponse(
+                  req,
+                  id.getOrElse(""),
+                  interpreter,
+                  skipValidation,
+                  enableIntrospection,
+                  queryExecution,
+                  subscriptions
+                ).catchAll(toStreamError(id, _))
+              case None      => connectionError
+            }
+          case GraphQLWSRequest("stop", id, _)                =>
+            removeSubscription(id, subscriptions)
+              .flatMap(_ => ZStream.empty)
+
+        })
+        .flatten
+        .catchAll(_ => connectionError)
```

Comment:
Can we send the error message (I'm comparing the implementation with the http4s one so for all my comments you can refer to it to see how it was done)?

Reply:
I think this is correct (at least according to https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md#gql_connection_error).

For errors during query execution we already respond with a `GQL_ERROR` message (L106), and all other cases should be bad input data from the client which should result in `GQL_CONNECTION_ERROR`. But I might be missing something!

Reply:
Looks good!